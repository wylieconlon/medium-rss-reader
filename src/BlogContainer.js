import React, { Component } from 'react';
import memoize from 'memoize-one';

import Article from './Article';

class BlogContainer extends Component {
  constructor(props) {
    super(props); 

    this.state = {
      expandedArticleKey: null,
    };

    // Prevents unnecessary re-renders as long as the document is the same
    this._processRSSDoc = memoize(this._getBlogInfo);
  }

  render() {
    const blogInfo = this._processRSSDoc(this.props.rss);

    if (blogInfo) {
      return <section>
        <div className="blogTitle">
          <h1>{blogInfo.title}</h1>
          <p dangerouslySetInnerHTML={blogInfo.description} />
        </div>

        {blogInfo.articles.map((article) => {
          const isExpanded = article.shortLink === this.state.expandedArticleKey;

          return (<Article key={article.shortLink}
            article={article}
            isExpanded={isExpanded}
            onExpand={this.expandArticle.bind(this, article)}
          />);
        })}
      </section>;
    } else {
      return <div></div>;
    }
  }

  expandArticle(article) {
    this.setState({ expandedArticleKey: article.shortLink });
  }

  _getBlogInfo() {
    const rssDocument = this.props.rss;
    if (rssDocument) {
      return {
        title: this._getSingleValue(rssDocument, 'channel > title'),
        description: {__html: this._getSingleValue(rssDocument, 'channel > description')},
        link: this._getSingleValue(rssDocument, 'channel > link'),
        imageUrl: this._getSingleValue(rssDocument, 'channel > image > url'),
        imageLink: this._getSingleValue(rssDocument, 'channel > image > link'),
        imageTitle: this._getSingleValue(rssDocument, 'channel > image > title'),
        articles: this._getBlogArticles(rssDocument),
      };
    }
  }

  _getBlogArticles(rssDocument) {
    if (rssDocument) {
      const articles = Array.from(rssDocument.querySelectorAll('item'));
      return articles.map((articleNode) => {
        const publicationDateString = this._getSingleValue(articleNode, 'pubDate');
        const publicationDateFormatted = new Date(publicationDateString).toLocaleDateString();
        const updatedDateString = this._getSingleValue(articleNode, 'updated');
        const updatedDateFormatted = new Date(updatedDateString).toLocaleDateString();
        const fullArticle = this._getSingleValue(articleNode, 'encoded');
        return {
          title: this._getSingleValue(articleNode, 'title'),
          description: {__html: this._getSingleValue(articleNode, 'description')},
          hasFullArticle: !!fullArticle,
          content: {__html: fullArticle},
          link: this._getSingleValue(articleNode, 'link'),
          categories: this._getMultiValue(articleNode, 'category'),
          shortLink: this._getSingleValue(articleNode, 'guid'),
          creator: this._getSingleValue(articleNode, 'creator'),
          publicationDateString,
          updatedDateString,
          publicationDateFormatted,
          updatedDateFormatted,
        };
      });
    }
  }

  _getSingleValue(rssDocument, selector) {
    const element = rssDocument.querySelector(selector);
    return element ? element.textContent : '';
  }

  _getMultiValue(rssDocument, selector) {
    return Array.from(rssDocument.querySelectorAll(selector))
      .map((node) => node.textContent);
  }
}

export default BlogContainer;