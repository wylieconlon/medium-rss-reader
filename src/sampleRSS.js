export default `
<?xml version="1.0" encoding="UTF-8"?><rss xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:content="http://purl.org/rss/1.0/modules/content/" xmlns:atom="http://www.w3.org/2005/Atom" version="2.0" xmlns:cc="http://cyber.law.harvard.edu/rss/creativeCommonsRssModule.html">
    <channel>
        <title><![CDATA[Blog Name - Medium]]></title>
        <description><![CDATA[Blog Description - Medium]]></description>
        <link>https://medium.com/blog-url</link>
        <image>
            <url>https://cdn-images-1.medium.com/proxy/1*TGH72Nnw24QL3iV9IOm4VA.png</url>
            <title>Compass True North - Medium</title>
            <link>https://medium.com/compass-true-north?source=rss----2a2f97cb73e3---4</link>
        </image>
        <generator>Medium</generator>
        <lastBuildDate>Wed, 17 Oct 2018 15:43:41 GMT</lastBuildDate>
        <atom:link href="https://medium.com/feed/compass-true-north" rel="self" type="application/rss+xml"/>
        <webMaster><![CDATA[yourfriends@medium.com]]></webMaster>
        <atom:link href="http://medium.superfeedr.com" rel="hub"/>
        <item>
            <title><![CDATA[Working with AWS & Agent Feedback to build a better product]]></title>
            <link>https://medium.com/compass-true-north/working-with-aws-agent-feedback-to-build-a-better-product-c3009b6e6184?source=rss----2a2f97cb73e3---4</link>
            <guid isPermaLink="false">https://medium.com/p/c3009b6e6184</guid>
            <category><![CDATA[product]]></category>
            <category><![CDATA[expansion]]></category>
            <category><![CDATA[compass]]></category>
            <category><![CDATA[engineering]]></category>
            <dc:creator><![CDATA[Brian Renzenbrink]]></dc:creator>
            <pubDate>Mon, 08 Oct 2018 14:55:35 GMT</pubDate>
            <atom:updated>2018-10-08T14:55:35.494Z</atom:updated>
            <content:encoded><![CDATA[<p>A few weeks back, we got to work with the folks at <a href="https://uncubed.com/">UnCubed</a> in partnership with AWS to put together a short video describing one of the early challenges in our expansion and how our infrastructure in AWS helped us work through it. Check itÂ out!</p>]]></content:encoded>
        </item>
    </channel>
</rss>
`;
