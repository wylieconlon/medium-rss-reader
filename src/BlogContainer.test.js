import React from 'react';
import { shallow } from 'enzyme';

import BlogContainer from './BlogContainer';
import sampleRSS from './sampleRSS';

it('renders empty without crashing', () => {
  shallow(<BlogContainer />);
});

it('renders a container from an RSS feed', () => {
  const rssDocument = new DOMParser().parseFromString(sampleRSS, 'application/xml');
  expect(shallow(<BlogContainer rss={rssDocument} />).html()).toMatchSnapshot();
});
