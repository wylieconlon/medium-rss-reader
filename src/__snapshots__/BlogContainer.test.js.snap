// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a container from an RSS feed 1`] = `"<section><div class=\\"blogTitle\\"><h1>Blog Name - Medium</h1><p>Blog Description - Medium</p></div><article class=\\"\\"><h2><a href=\\"https://medium.com/compass-true-north/working-with-aws-agent-feedback-to-build-a-better-product-c3009b6e6184?source=rss----2a2f97cb73e3---4\\">Working with AWS &amp; Agent Feedback to build a better product</a></h2><p>By </p><p></p><footer>Published <time dateTime=\\"Mon, 08 Oct 2018 14:55:35 GMT\\">2018-10-8</time>  • Updated at <time dateTime=\\"\\">Invalid Date</time></footer></article></section>"`;
